File: lib/src/main/java/vproxy/vserver/SubPath.java
package vproxy.vserver;

import vproxy.vserver.route.FixedSubPath;
import vproxy.vserver.route.VariableSubPath;
import vproxy.vserver.route.WildcardSubPath;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public interface SubPath {
    static SubPath create(String path) {
        List<String> paths = Arrays.stream(path.split("/")).map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList());
        SubPath next = null;
        for (int i = paths.size() - 1; i >= 0; --i) {
            String p = paths.get(i);
            if (p.equals("*")) {
                next = new WildcardSubPath(next);
            } else if (p.startsWith(":")) {
                next = new VariableSubPath(next, p.substring(1));
            } else {
                next = new FixedSubPath(next, p);
            }
        }
        return next;
    }

    SubPath next();

    boolean match(String route);

    boolean currentSame(SubPath r);

    void fill(RoutingContext ctx, String route);
}

File: lib/src/main/java/io/vproxy/lib/http/route/SubPath.kt
package io.vproxy.lib.http.route

import io.vproxy.lib.http.RoutingContext
import java.util.*
import java.util.stream.Collectors

interface SubPath {
  operator fun next(): SubPath?
  fun match(route: String): Boolean
  fun currentSame(r: SubPath): Boolean
  fun fill(ctx: RoutingContext, route: String)

  companion object {
    @JvmStatic
    fun create(path: String): SubPath? {
      val paths = Arrays.stream(path.split("/").toTypedArray()).map { obj: String -> obj.trim { it <= ' ' } }
        .filter { s: String -> !s.isEmpty() }.collect(Collectors.toList())
      var next: SubPath? = null
      for (i in paths.indices.reversed()) {
        val p = paths[i]
        next = if (p == "*") {
          WildcardSubPath(next)
        } else if (p.startsWith(":")) {
          VariableSubPath(next, p.substring(1))
        } else {
          FixedSubPath(next, p)
        }
      }
      return next
    }
  }
}
